File: examples/rust/examples/ttl_expiration_test.rs
use ekodb_client::{Client, FieldType, Record};
use std::env;
use std::thread;
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("╔════════════════════════════════════════════════════════╗");
    println!("║     TTL EXPIRATION VERIFICATION TEST                   ║");
    println!("╚════════════════════════════════════════════════════════╝");
    println!();
    println!("This test verifies that document TTL expiration works correctly.");
    println!("We will insert documents with short TTL and verify they expire.");
    println!();

    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;
    println!("✓ Client connected");

    let collection = "ttl_expiration_test_rs";
    let ttl_seconds = 3;

    println!();
    println!("═══════════════════════════════════════════════════════════");
    println!("TEST 1: Document TTL Expiration");
    println!("═══════════════════════════════════════════════════════════");

    println!("\n[Step 1] Insert document with {} second TTL", ttl_seconds);
    println!("  Input: {{name: 'TTL Test', value: 'should expire'}}");
    println!("  TTL: {}s", ttl_seconds);

    let mut record = Record::new();
    record.insert("name", "TTL Test");
    record.insert("value", "should expire");
    let record = record.with_ttl(&format!("{}s", ttl_seconds));

    let doc = client.insert(collection, record, None).await?;
    let doc_id = match doc.get("id") {
        Some(FieldType::String(s)) => s.clone(),
        _ => return Err("No ID returned".into()),
    };
    println!("  Output: Document ID = {}", doc_id);
    println!("  ✓ PASS: Document inserted");

    println!("\n[Step 2] Verify document exists immediately");
    println!("  Input: find_by_id({})", doc_id);

    let found = client.find_by_id(collection, &doc_id, None).await?;
    let name_val = match found.fields.get("name") {
        Some(FieldType::String(s)) => s.as_str(),
        _ => "unknown",
    };
    println!("  Output: Found document with name = {}", name_val);
    println!("  ✓ PASS: Document exists");

    let wait_time = ttl_seconds + 2;
    println!("\n[Step 3] Wait for TTL to expire ({}s)", wait_time);
    print!("  Waiting");
    for _ in 0..wait_time {
        thread::sleep(Duration::from_secs(1));
        print!(".");
        std::io::Write::flush(&mut std::io::stdout())?;
    }
    println!(" done");
    println!("  ✓ PASS: Wait complete");

    println!("\n[Step 4] Verify document has expired");
    println!("  Input: find_by_id({})", doc_id);

    match client.find_by_id(collection, &doc_id, None).await {
        Ok(_) => {
            println!("  Output: Document still exists!");
            return Err("❌ FAILED: Document should have expired but still exists!".into());
        }
        Err(e) => {
            println!("  Output: Error (expected) - {}", e);
            println!("  ✓ PASS: Document expired (not found error)");
        }
    }

    println!("\n═══════════════════════════════════════════════════════════");
    println!("CLEANUP");
    println!("═══════════════════════════════════════════════════════════");

    if let Err(e) = client.delete_collection(collection).await {
        println!("Warning: cleanup failed: {}", e);
    } else {
        println!("✓ Deleted test collection");
    }

    println!("\n╔════════════════════════════════════════════════════════╗");
    println!("║              ALL TESTS PASSED ✓                        ║");
    println!("╚════════════════════════════════════════════════════════╝");

    Ok(())
}

File: examples/typescript/ttl_expiration_test.ts
/**
 * TTL Expiration Verification Test
 *
 * This test VERIFIES that TTL actually works by:
 * 1. Inserting a document with short TTL (3 seconds)
 * 2. Verifying the document exists immediately
 * 3. Waiting for TTL to expire
 * 4. Verifying the document is GONE
 *
 * If this test passes, TTL expiration is working correctly.
 */

import { EkoDBClient } from "@ekodb/ekodb-client";
import * as dotenv from "dotenv";

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  console.log("╔════════════════════════════════════════════════════════╗");
  console.log("║     TTL EXPIRATION VERIFICATION TEST                   ║");
  console.log("╚════════════════════════════════════════════════════════╝");
  console.log();
  console.log(
    "This test verifies that document TTL expiration works correctly.",
  );
  console.log(
    "We will insert documents with short TTL and verify they expire.",
  );
  console.log();

  const client = new EkoDBClient(BASE_URL, API_KEY);
  await client.init();
  console.log("✓ Client connected");

  const collection = "ttl_expiration_test_ts";
  const ttlSeconds = 3;

  console.log();
  console.log("═══════════════════════════════════════════════════════════");
  console.log("TEST 1: Document TTL Expiration");
  console.log("═══════════════════════════════════════════════════════════");

  console.log(`\n[Step 1] Insert document with ${ttlSeconds} second TTL`);
  console.log("  Input: {name: 'TTL Test', value: 'should expire'}");
  console.log(`  TTL: ${ttlSeconds}s`);

  const testDoc = {
    name: "TTL Test",
    value: "should expire",
  };

  const insertedDoc = await client.insert(collection, testDoc, {
    ttl: ttlSeconds + "s",
  });

  const docID = insertedDoc.id as string;
  console.log("  Output: Document ID = " + docID);
  console.log("  ✓ PASS: Document inserted");

  console.log("\n[Step 2] Verify document exists immediately");
  console.log("  Input: findById(" + docID + ")");

  const found = await client.findById(collection, docID);
  if (!found) {
    throw new Error("❌ FAILED: Document should exist but was not found");
  }
  const foundName = (found.name as any)?.value || found.name;
  console.log("  Output: Found document with name = " + foundName);
  console.log("  ✓ PASS: Document exists");

  const waitTime = (ttlSeconds + 2) * 1000;
  console.log("\n[Step 3] Wait for TTL to expire (" + waitTime / 1000 + "s)");
  process.stdout.write("  Waiting");
  for (let i = 0; i < waitTime / 1000; i++) {
    await sleep(1000);
    process.stdout.write(".");
  }
  console.log(" done");
  console.log("  ✓ PASS: Wait complete");

  console.log("\n[Step 4] Verify document has expired");
  console.log("  Input: findById(" + docID + ")");

  try {
    const expired = await client.findById(collection, docID);
    if (expired === null || expired === undefined) {
      console.log("  Output: null (document not found)");
      console.log("  ✓ PASS: Document expired correctly!");
    } else {
      console.log(
        "  Output: Document still exists! " + JSON.stringify(expired),
      );
      throw new Error(
        "❌ FAILED: Document should have expired but still exists!",
      );
    }
  } catch (err: any) {
    if (err.message.includes("FAILED")) throw err;
    console.log("  Output: Error (expected) - " + err.message);
    console.log("  ✓ PASS: Document expired (not found error)");
  }

  console.log("\n═══════════════════════════════════════════════════════════");
  console.log("CLEANUP");
  console.log("═══════════════════════════════════════════════════════════");

  await client.deleteCollection(collection);
  console.log("✓ Deleted test collection");

  console.log();
  console.log("╔════════════════════════════════════════════════════════╗");
  console.log("║              ALL TESTS PASSED ✓                        ║");
  console.log("╚════════════════════════════════════════════════════════╝");
}

main().catch((error) => {
  console.error("❌ Error:", error);
  process.exit(1);
});
